### PR TITLE
[PA-568] Adding deeper validation in Backup Operation

### PR DESCRIPTION
**Pull request by tthurlapati-px**
## What this PR does

1. It adds the ability to Validate the finer details of a backup (to ensure that it has been done correctly). This includes whether all specs are backed up, all volumes and namespaces are with all their properties.
2. We can now filter out the backedup specs (which can by used to later validate restore, and even clean up restored namespaces)
4. Updated `BasicBackupCreation` to use a "standardized" method of writing code that obeys invariants that are built into my implementation of deeper backup validation such that there is lesser chance of writing code that can fail.
5. `GetUpdatedSpec` will now help get the latest spec.
6. A workaround has been added, where the kind value is manually added :(
7. Adapt `UpdateNamespace` to use empty namespace mappings
8. ADD `GetAppStorageClasses` which can get the storage classes associated with the PVCs in a namespace
9. MODIFY `GetSpecNameKindNamepace` to use function calls rather than direct variable
10. `FilterAppContextsByNamespace` is used to get filter a bunch of schedule app contexts to only include those with the specified namespaces.
11. `CreateBackupWithValidation` creates backup and validates
12. `CreateBackupWithCustomResourceTypeWithValidation` creates backup and validates incl custom resourrce type specification

### Tests using `ValidateBackup`

- [x] BasicBackupCreation

## Which issue(s) this PR fixes
Closes #PA-568

## Special notes for your reviewer

The links to the Jenkins run are further down in comments.

## HACK

The `kind` field missing on the objects returned by certain sched-ops functions (that are used in `k8s.go` seems to be an issue with kubernetes itself:
https://github.com/kubernetes/kubernetes/issues/80609
https://github.com/kubernetes/kubernetes/issues/3030

Currently, this issue seems to be affecting those calls in sched-ops which get a list and then filter out the required object, like the following example where Services and fetched and service is filtered: `c.kubernetes.CoreV1().Services(ns).Create(context.TODO(), service, metav1.CreateOptions{})`
That's the reason some of them don't have kind (list being fetched causes apiVersion and kind to be dropped due to k8s bug where TypeMeta is not found, I think), whereas some do have kind.

Basically, in `k8s.go` this is what is happening:
```go
else if obj, ok := spec.(*corev1.Service); ok {
		obj.Namespace = ns.Name
// obj.Kind is PRESENT as per logs
		svc, err := k8sCore.CreateService(obj) //this could be GetService(...)
		if err != nil {
			return nil, &scheduler.ErrFailedToScheduleApp{
				App:   app,
				Cause: fmt.Sprintf("Failed to create Service: %v. Err: %v", obj.Name, err),
			}
		}
// svc.Kind is EMPTY as per logs
		// This is a hack, because the `Kind` is empty for some reason
		svc.Kind = "Service"
// now svc.Kind is PRESENT as per logs
		log.Infof("[%v] Created Service: %v", app.Key, svc.Name)
		return svc, nil

	} 
```

So, as a solution, as seen above, I am manually assigning the `Kind` field. When the K8s bug gets fixed, we don't need to do that any longer.

### CODE REVIEW

**Review by mkoppal-px — CHANGES_REQUESTED** [review]
I am not very convinced with the way `ValidateBackup` function is written. We need to optimize the number of inputs it takes.
Let us discuss this.

File: tests/backup/backup_helper.go [hunk 1]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+// * An error can be returned *with* backedupAppContexts, so do check if the first variable returned is actually nil
+// *
+// * # Parameters
+// *
+// * requireAllScheduledCtxAreInBackup: This requires that Backup objects are a superset of objects in ScheduledCtx: If set to true, error is returned if something in scheduledCtx is not present in BackupCtx (set to false if not everything in scheduledCtx has been backed up, on purpose)
+// * requireAllBackupAreInSomeScheduledCtx: This requires that ScheduledCtx objects are a superset of objects in Backup: If set to true, error is returned if something in Backup is not present in scheduledCtx. (set to false if deploying CRs)
+// * NOTES: Setting both to true will check for strict equality of both sets. Set both to false if you just want to filter and get whatever is possible.
+// *
+// * NOTE: Ensure scheduledCtxs which correspond to the namespaces backed up in the backup are provided
+// * NOTE: This function must be called after the backup is completed (with status Success/PartialSuccess)
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1259]: Reduce the comments here and make it crisp

File: tests/backup/backup_helper.go [hunk 2]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+func ValidateBackup(ctx context.Context, backupName string, orgID string, scheduledAppContexts []*scheduler.Context, requireAllScheduledCtxAreInBackup, requireAllBackupAreInSomeScheduledCtx bool, backupClusterConfigPath string) ([]*scheduler.Context, error) {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1260]: Remove the parameters - `requireAllScheduledCtxAreInBackup, requireAllBackupAreInSomeScheduledCtx bool, backupClusterConfigPath string`

File: tests/backup/backup_helper.go [hunk 3]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+		return nil, fmt.Errorf("InspectBackup Err: %v", err)
+	}
+
+	backupStatus := backupInspectResponse.GetBackup().GetStatus().Status
+	if backupStatus != api.BackupInfo_StatusInfo_Success &&
+		backupStatus != api.BackupInfo_StatusInfo_PartialSuccess {
+		return nil, fmt.Errorf("ValidateBackup requires backup [%s] to have a status of Success or PartialSuccess", backupName)
+	}
+
+	backedupAppContexts, ScheduledCtxNotFoundInBackupErrors, BackupNotFoundInAnyScheduledCtxErrors, otherErrors := GetBackupCtxsFromScheduledCtxs(backupInspectResponse, scheduledAppContexts)
```
[the remark this answers is not in the document]
Comment (mkoppal-px) [reply tests/backup/backup_helper.go:1285]: We have agreed that we will be validating whatever is there as part of the spec should be in the backup inspect response.

File: tests/backup/backup_helper.go [hunk 4]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+}
+
+// ValidateBackedUpVolumes checks if the volumes have been backed up and done so rightly
+// *
+// * Returns:
+// * otherErrors: all the other kind of error; Don't ignore these are they're serious errors
+// *
+// * NOTES:
+// * - make sure to include Contexts of *all* namespaces which are supposed to contain the volumes
+func ValidateBackedUpVolumes(backupInspectResponse *api.BackupInspectResponse,
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1322]: As part of volume validation, we will just check that volumes given by k8s API is present in the backup inspect response

File: tests/backup/backup_helper.go [hunk 5]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+	scheduledAppContexts []*scheduler.Context) (errors []error) {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1323]: Consolidate all the errors happening in this function and return a single error

File: tests/backup/backup_helper.go [hunk 6]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+	if backupStatus != api.BackupInfo_StatusInfo_Success &&
+		backupStatus != api.BackupInfo_StatusInfo_PartialSuccess {
+		return nil, fmt.Errorf("ValidateBackup requires backup [%s] to have a status of Success or PartialSuccess", backupName)
+	}
+
+	backedupAppContexts, ScheduledCtxNotFoundInBackupErrors, BackupNotFoundInAnyScheduledCtxErrors, otherErrors := GetBackupCtxsFromScheduledCtxs(backupInspectResponse, scheduledAppContexts)
+	// print all error, but don't consume them
+	ProcessMultipleErrors("GetBackupCtxsFromScheduledCtxs-ScheduledCtxNotFoundInBackupErrors", ScheduledCtxNotFoundInBackupErrors, true)
+	ProcessMultipleErrors("GetBackupCtxsFromScheduledCtxs-BackupNotFoundInAnyScheduledCtxErrors", BackupNotFoundInAnyScheduledCtxErrors, true)
+	ProcessMultipleErrors("GetBackupCtxsFromScheduledCtxs-otherErrors", otherErrors, true)
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1289]: As discussed, we may not need this for now. Just consolidate the errors.

File: tests/backup/backup_helper.go (hunk 2 above)
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1260]: Remove returning `[]*scheduler.Context`

File: drivers/scheduler/k8s/k8s.go [hunk 7]
```diff
@@ -1317,6 +1317,23 @@ func (k *K8s) UpdateTasksID(ctx *scheduler.Context, id string) error {
 	return nil
 }
 
+func (k *K8s) GetUpdatedSpec(spec interface{}) (interface{}, error) {
```
Comment (lsrinivas-pure) [inline drivers/scheduler/k8s/k8s.go:1320]: add doc string for this

File: drivers/scheduler/k8s/k8s.go [hunk 8]
```diff
@@ -1317,6 +1317,23 @@ func (k *K8s) UpdateTasksID(ctx *scheduler.Context, id string) error {
...
+	if obj, ok := spec.(*corev1.PersistentVolumeClaim); ok {
+		pvc, err := k8sCore.GetPersistentVolumeClaim(obj.Name, obj.Namespace)
+
+		if err != nil {
+			return nil, err
+		}
+
+		// This is a hack, because the `Kind` is empty for some reason
+		pvc.Kind = "PersistentVolumeClaim"
```
Comment (lsrinivas-pure) [inline drivers/scheduler/k8s/k8s.go:1329]: can you give any sample response of complete pvc struct including TypeMeta

File: tests/backup/backup_helper.go (hunk 2 above)
Comment (lsrinivas-pure) [reply tests/backup/backup_helper.go:1260]: too many parameters is not, see if you can reduce or pass some backup struct will all the fields

File: tests/backup/backup_helper.go [hunk 9]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+// * Returns:
+// * backupclusterAppsContexts: the filtered context (backup)
+// * ScheduledCtxNotFoundInBackupErrors: these are errors that are generated when namespaces and/or specObjs in the scheduled context are not found in the backup
+// * BackupNotFoundInAnyScheduledCtxErrors: these are errors that are generated when namespaces and/or resources in the backup context are not found in the scheduled contextx
+// * otherErrors: all the other kinds of errors; Don't ignore these are they're serious errors
+// *
+// * NOTES:
+// * - make sure to include Contexts of *all* namespaces which are supposed to contain the backup objects
+// * - The errors returned are all tolerable errors that have been caught until an intolerable error was encountered
+func GetBackupCtxsFromScheduledCtxs(backupInspectResponse *api.BackupInspectResponse, scheduledAppContexts []*scheduler.Context) (backupclusterAppsContexts []*scheduler.Context, ScheduledCtxNotFoundInBackupErrors, BackupNotFoundInAnyScheduledCtxErrors, otherErrors []error) {
```
Comment (lsrinivas-pure) [inline tests/backup/backup_helper.go:1471]: too many return args just for errors. just return error if there is an error. keep it simple

File: tests/backup/backup_helper.go [hunk 10]
```diff
@@ -1242,7 +1246,377 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+	errGetBackupCtxsFromScheduledCtxs := ProcessMultipleErrors("GetBackupCtxsFromScheduledCtxs", errArr, false)
+
+	errors := ValidateBackedUpVolumes(backupInspectResponse, scheduledAppContexts)
+	errValidateBackedUpVolumes := ProcessMultipleErrors("ValidateBackedUpVolumes", errors, true)
+
+	if errGetBackupCtxsFromScheduledCtxs != nil || errValidateBackedUpVolumes != nil {
+		err = fmt.Errorf("GetBackupCtxsFromScheduledCtxs: %v ;; ValidateBackedUpVolumes: %v",
+			errGetBackupCtxsFromScheduledCtxs, errValidateBackedUpVolumes)
+	} else {
+		err = nil
```
Comment (lsrinivas-pure) [inline tests/backup/backup_helper.go:1309]: I dont think this is required, please have a look

File: tests/backup/backup_helper.go [hunk 11]
```diff
@@ -1322,6 +1696,15 @@ func restoreSuccessWithReplacePolicy(restoreName string, orgID string, retryDura
 	return err
 }
 
+// ValidateRestore returns a clone of the contexts (with backup objects) *after* converting the contexts to point to restored objects (and after validating those objects)
+func ValidateRestore(ctx context.Context, restoreName string, orgID string, backedupAppContexts []*scheduler.Context, namespaceMapping map[string]string, storageClassMapping map[string]string, restoreClusterConfigPath string) ([]*scheduler.Context, error) {
```
Comment (lsrinivas-pure) [inline tests/backup/backup_helper.go:1700]: please see if you minimise the args

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

**Comment by tthurlapati-px** [comment]
@lsrinivas-pure  I will address all the comments you made later on, as based on my discussion with Mithun, I am making some of the code simpler and easier by not covering edge cases which aren't needed and will make things unnecessarily complicated.

Though, I _am_ addressing the `kind` field is missing issue as that is of paramount importance. Here is an example of a `PersistentVolumeClaim` object printed with `%+v`:
```
&PersistentVolumeClaim{ObjectMeta:{mysql-data mysql-backup-pxbackuptask-0-05-01-09h33m27s 97d5f87f-6e50-48f6-a9d5-384aaefadd7d 7798919 0 2023-05-01 09:35:30 +0000 UTC map[] map[px/secret-key:mysql-secret px/secret-name:volume-secrets px/secret-namespace:mysql-backup-pxbackuptask-0-05-01-09h33m27s] [] [kubernetes.io/pvc-protection] [{backup.test Update v1 2023-05-01 09:35:30 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:px/secret-key":{},"f:px/secret-name":{},"f:px/secret-namespace":{}}},"f:spec":{"f:accessModes":{},"f:resources":{"f:requests":{".":{},"f:storage":{}}},"f:storageClassName":{},"f:volumeMode":{}}} }]},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{2147483648 0} {} 2Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*mysql-sc,VolumeMode:*Filesystem,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:Pending,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},ResizeStatus:nil,},}
```
ref: https://aetos.pwx.purestorage.com/resultSet/testSetID/177665/testCaseID/914767

There IS no `TypeMeta`. Other stuff is there.

File: tests/backup/backup_helper.go (hunk 11 above)
Comment (tthurlapati-px) [reply tests/backup/backup_helper.go:1700]: No, all those are needed. But this function will be handled in next PR anyway. we can resolve issues then...this is placeholder for demo

File: drivers/scheduler/k8s/k8s.go (hunk 7 above)
Comment (tthurlapati-px) [reply drivers/scheduler/k8s/k8s.go:1320]: ✅

**Comment by tthurlapati-px** [comment]
@mkoppal-px  I have made all the changes you requested.

File: tests/backup/backup_helper.go [hunk 12]
```diff
@@ -1245,7 +1263,202 @@ func backupSuccessCheck(backupName string, orgID string, retryDuration time.Dura
...
+		Name:  backupName,
+		Uid:   backupUid,
+		OrgId: orgID,
+	}
+	backupInspectResponse, err := backupDriver.InspectBackup(ctx, backupInspectRequest)
+	if err != nil {
+		return fmt.Errorf("InspectBackup Err: %v", err)
+	}
+
+	backupStatus := backupInspectResponse.GetBackup().GetStatus().Status
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1286]: I would suggest to declare a variable and assign it as follows to avoid repeating this multiple times `backupInspectResponse.GetBackup()`

```
backupObj := backupInspectResponse.GetBackup()
```

**Comment by mkoppal-px** [comment]
Once all the comments are addressed, please run a few tests where you are using `ValidateBackup` after `CreateBackup` is called.

File: tests/backup/backup_helper.go (hunk 12 above)
Comment (tthurlapati-px) [reply tests/backup/backup_helper.go:1286]: ✅

**Comment by tthurlapati-px** [comment]
@mkoppal-px  the runs of `DeleteAllBackupObjects,BasicSelectiveRestore,CustomResourceBackupAndRestore,BasicBackupCreation`:

mysql:
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/1279/consoleFull
https://aetos.pwx.purestorage.com/resultSet/testSetID/179841

mysql: Running `BasicBackupCreation` again, as it failed because of PVs not getting created or something.
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/1281/consoleFull
https://aetos.pwx.purestorage.com/resultSet/testSetID/179912

postgres:
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/1283/consoleFull
https://aetos.pwx.purestorage.com/resultSet/testSetID/179931

File: tests/backup/backup_helper.go [hunk 13]
```diff
@@ -201,6 +204,30 @@ func CreateBackup(backupName string, clusterName string, bLocation string, bLoca
...
+	for _, appContext := range appContexts {
+		if Contains(namespaces, appContext.ScheduleOptions.Namespace) {
+			filteredAppContexts = append(filteredAppContexts, appContext)
+		}
+	}
+	return
+}
+
+// CreateBackupWithValidatation creates backup, checks for success, and validates the backup
+func CreateBackupWithValidatation(ctx context.Context, backupName string, clusterName string, bLocation string, bLocationUID string, scheduledAppContexts []*scheduler.Context, labelSelectors map[string]string, orgID string, uid string, preRuleName string, preRuleUid string, postRuleName string, postRuleUid string) error {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:218]: nit: It should be `CreateBackupWithValidation`

**Comment by mkoppal-px** [comment]
Since we are confident now that `CreateBackupWithValidation` works with no problem. Can we run all the tests by replacing `CreateBackup` with `CreateBackupWithValidation` and use `postgres-backup` app?

File: tests/backup/backup_helper.go (hunk 13 above)
Comment (tthurlapati-px) [reply tests/backup/backup_helper.go:218]: 😅 fixed

**Comment by tthurlapati-px** [comment]
@mkoppal-px  which tests would you like to use it with? Only the ones in nightly, or _every_ test in PX-Backup Automation?

**Comment by mkoppal-px** [comment]
Every test which is using `CreateBackup` should now use `CreateBackupWithValidation`.
I would also want to know if it taking more time because of the Validation part.

**Comment by tthurlapati-px** [comment]
Alright.
As per the logs, **Backup Validation doesn't take more than a second**.
Validation of backup is just checking if the list of backed up objects given by PX-backup consists of whatever is already in the spec, (which means the only network call made is to get list of volumes for the spec), which explains why its fast enough for us.

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Comment by kshithijiyer-px** [comment]
@tthurlapati-px Can you update the description of the PR?

**Comment by mkoppal-px** [comment]
Revert all the changes to the tests except for `BasicBackupCreation` and then merge it. Using `CreateBackupWithValidation` for all the test can be taken up in a separate PR as per our discussion.

**Comment by tthurlapati-px** [comment]
@mkoppal-px  done

File: tests/backup/backup_restore_basic_test.go [hunk 14]
```diff
@@ -108,6 +108,7 @@ var _ = Describe("{BasicSelectiveRestore}", func() {
 		})
 		Step("Taking backup of multiple namespaces", func() {
 			log.InfoD(fmt.Sprintf("Taking backup of multiple namespaces [%v]", bkpNamespaces))
+
```
Comment (mkoppal-px) [inline tests/backup/backup_restore_basic_test.go:111]: nit: Remove this new line

**Review by mkoppal-px — APPROVED** [review]
LGTM except for one minor comment
